File: 백준/Silver/9012. 괄호/괄호.py
# 9012 괄호
n = int(input())

for i in range(n):
    stack = []
    a = list(input())

    for j in range(len(a)):
        if not stack:
            stack.append(a[j])
        elif stack[-1]=="(" and a[j]==')':
            stack.pop()
        else:
            stack.append(a[j])

    if not stack:
        print('YES')
    else:
        print('NO')

File: 백준/Silver/9012. 괄호/괄호.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        for (int i = 0; i<n; i++) {
            String s = br.readLine();
            Stack<Character> stack = new Stack<>();
            for (int j = 0; j<s.length(); j++) {
                if (s.charAt(j)=='(') {
                    stack.push(s.charAt(j));
                }
                else {
                    if (stack.empty()) {
                        stack.push(s.charAt(j));
                        break;
                    } else {
                        stack.pop();
                    }
                }
            }
            if (stack.empty()) {
                System.out.println("YES");
            } else {
                System.out.println("NO");
            }
        }
    }
}